src/PrestaShopBundle/Controller/Admin/Sell/Order/CartController.php

```diff
@@ -55,14 +60,142 @@
 use PrestaShop\PrestaShop\Core\Domain\Product\Exception\PackOutOfStockException;
 use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductCustomizationNotFoundException;
 use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductOutOfStockException;
+use PrestaShop\PrestaShop\Core\Search\Filters\CartFilter;
+use PrestaShopBundle\Component\CsvResponse;
 use PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController;
+use PrestaShopBundle\Controller\BulkActionsTrait;
 use PrestaShopBundle\Security\Annotation\AdminSecurity;
 use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\RedirectResponse;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 
 class CartController extends FrameworkBundleAdminController
 {
+    use BulkActionsTrait;
+
+    /**
+     * Shows list of carts
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
+     *
+     * @param Request $request
+     * @param CartFilter $filters
+     *
+     * @return Response
+     */
+    public function indexAction(Request $request, CartFilter $filters): Response
+    {
+        $cartsKpiFactory = $this->get('prestashop.core.kpi_row.factory.carts');
+        $cartGridFactory = $this->get('prestashop.core.grid.factory.cart');
```

Comment:
Use DI instead

Reply:
Sadly we don't really have proper ways to inject these services with DI, there would be alternatives though:
- injecting in the constructor, but I wouldn't recommend it because not all actions need all these services (grid and form are used in different actions)
- the simplest one is to add these service in `getSubscribedServices`
- we'd need a public factory that allows building the grid, for builder and form handler but it requires some refacto

So I agree this is not ideal, but since we didn't invest time on how to improve our controllers we have no alternative ready to use, and most of our migrated pages are the same so accepting this page as is is at least one more step towards migration, until we can refacto it properly along with all the other controllers.

Reply:
I understand. For information, this is normally no longer possible in SF6, we have kept a compatibility layer so that it continues to work, but the day we switch to a "clean" implementation, it will no longer work.